packages/xforms-engine/src/instance/RankControl.ts

```diff
@@ -0,0 +1,144 @@
+import { XPathNodeKindKey } from '@getodk/xpath';
+import type { Accessor } from 'solid-js';
+import { createMemo } from 'solid-js';
+import type {
+	RankDefinition,
+	RankItem,
+	RankNode,
+	RankValueOptions,
+} from '../client/RankNode.ts';
+import type { TextRange } from '../client/TextRange.ts';
+import type { XFormsXPathElement } from '../integration/xpath/adapter/XFormsXPathNode.ts';
+import type { CurrentState } from '../lib/reactivity/node-state/createCurrentState.ts';
+import type { EngineState } from '../lib/reactivity/node-state/createEngineState.ts';
+import type { SharedNodeState } from '../lib/reactivity/node-state/createSharedNodeState.ts';
+import { createSharedNodeState } from '../lib/reactivity/node-state/createSharedNodeState.ts';
+import { createFieldHint } from '../lib/reactivity/text/createFieldHint.ts';
+import { createNodeLabel } from '../lib/reactivity/text/createNodeLabel.ts';
+import type { SimpleAtomicState } from '../lib/reactivity/types.ts';
+import type { Root } from './Root.ts';
+import type { ValueNodeStateSpec } from './abstract/ValueNode.ts';
+import { ValueNode } from './abstract/ValueNode.ts';
+import type { GeneralParentNode } from './hierarchy.ts';
+import type { EvaluationContext } from './internal-api/EvaluationContext.ts';
+import type { ValidationContext } from './internal-api/ValidationContext.ts';
+import type { ClientReactiveSubmittableValueNode } from './internal-api/submission/ClientReactiveSubmittableValueNode.ts';
+import { RankFunctionalityError, RankValueTypeError } from '../error/RankError.ts';
+import { BaseItemCollectionCodec } from '../lib/codecs/BaseItemCollectionCodec.ts';
+import { sharedValueCodecs } from '../lib/codecs/getSharedValueCodec.ts';
+import type { AnyNodeDefinition } from '../parse/model/NodeDefinition.ts';
+import { createItemCollection } from '../lib/reactivity/createItemCollection.ts';
+
+type AssertRangeNodeDefinition = (definition: RankDefinition) => asserts definition is RankDefinition<'string'>;
+const assertRangeNodeDefinition: AssertRangeNodeDefinition = (definition) => {
+	if (definition.valueType !== 'string') {
+		throw new RankValueTypeError(definition);
+	}
+};
```

Comment:
```suggestion
export type AnyRankDefinition = {
	[V in ValueType]: RankDefinition<V>;
}[ValueType];

type AssertRankNodeDefinition = (
	definition: AnyRankDefinition
) => asserts definition is RankDefinition<'string'>;

const assertRankNodeDefinition: AssertRankNodeDefinition = (definition) => {
	if (definition.valueType !== 'string') {
		throw new RankValueTypeError(definition);
	}
};
```

1. Note the name changes (range -> rank). (I make similar-word typos like this _all the time_ and I know it's maddening! I can't count the number of times I've typed `NoteNote` or `NodeNode` 🙃)
2. Let me know if the use of mapped type for `AnyRankDefinition` isn't clear. The intent is to derive the type so you don't need to define a huge union just to discard most of it.

I'll add a bit more explanation at the `from` signature, and hopefully tie the whole definition type parameter -> `'string'` thing together.

Reply:
> Note the name changes (range -> rank)

🤦🏽‍♀️ Thanks!